docs/design/node_agent_config-design.md

```diff
@@ -0,0 +1,517 @@
+
+## High-Level Design
+
+Since new `nodeagentconfig` configuration option is a sibling of the `restic` one, the new common structure `NodeAgentCommonFields` will be created which will be exactly the same data structure as current `ResticConfig` and will be used by both `Restic` and the new `NodeAgentConfig`. The only difference between `NodeAgentConfig` and `Restic` is an addition of one `UploaderType` option to the `NodeAgentConfig` that will be either `kopia` or `restic`.
+
+Without the `UploaderType` option a default `restic` will be used. That default option may change in the future.
+
+## Detailed Design
+
+
+### New data structures
+
+A new structure will be added, that includes the ResticConfig inline and extends this with one new parameter `UploaderType`:
+
+```go
+type NodeAgentConfig struct {
+	// Embedding NodeAgentCommonFields
+	// +optional
+	NodeAgentCommonFields `json:",inline"`
+
+	// The type of uploader to transfer the data of pod volumes, the supported values are 'restic' or 'kopia'
+	// +optional
+	// +kubebuilder:validation:Enum=restic;kopia
+	UploaderType string `json:"uploaderType,omitempty"`
+}
+```
+
+The `NodeAgentCommonFields` structure is 1-1 as the current `Restic` structure.
+
+```go
+type NodeAgentCommonFields struct {
+	// enable defines a boolean pointer whether we want the daemonset to
+	// exist or not
+	// +optional
+	Enable *bool `json:"enable,omitempty"`
+	// supplementalGroups defines the linux groups to be applied to the Restic Pod
+	// +optional
+	SupplementalGroups []int64 `json:"supplementalGroups,omitempty"`
+	// timeout defines the Restic timeout, default value is 1h
+	// +optional
+	Timeout string `json:"timeout,omitempty"`
+	// Pod specific configuration
+	PodConfig *PodConfig `json:"podConfig,omitempty"`
+}
+```
+
+Current `Rescrit` structure will embedd the `NodeAgentCommonFields` without any additional options.
+
+The above `NodeAgentConfig` is a member of `ApplicationConfig`, which already includes `ResticConfig`, however we do not replace the `ResticConfig`, instead for backwards compatibility we add new `NodeAgentConfig` parameter:
+
+```go
+// ApplicationConfig defines the configuration for the Data Protection Application
+type ApplicationConfig struct {
+	Velero *VeleroConfig `json:"velero,omitempty"`
+	// (deprecation warning) ResticConfig is the configuration for restic server.
+	// Restic is for backwards compatibility and will be replaced by nodeagentconfig
+	// with the OADP 1.4+
+	// +optional
+	Restic *ResticConfig `json:"restic,omitempty"`
+
+	// NodeAgentConfig is needed to allow selection between kopia or restic
+	// +kubebuilder:validation:Optional
+	NodeAgent *NodeAgentConfig `json:"nodeAgentConfig,omitempty"`
+}
+```
+
+### Configuration YAML options
+
+The generated YAML manifest for the `nodeAgentConfig` and `restic` configuration options will be as follows:
+
+```yaml
+configuration:
```

Comment:
to avoid this big yaml, something like this would be valid?

> the updated DPA will look like this:
```yaml
kind: DataProtectionApplication
apiVersion: oadp.openshift.io/v1alpha1
metadata:
  ...
spec:
  ...
  configuration:
    ...
    restic:
      <Not changed, but added deprecated warning in description>
    nodeAgentConfig:
      uploaderType: <The type of uploader to transfer the data of pod volumes, the supported values are 'restic' or 'kopia'>
      <same fields restic has>
```

Reply:
Ok, but we are adding a new `nodeAgentConfig` which will contain all options that `restic` did +`uploaderType` so in fact entire `nodeAgentConfig` is what's changing. This way should we include entire part of the yaml for `nodeAgentConfig` ?

Reply:
I think there is no problem, just instead of going over every field of PodConfig (I thing is the biggest thing) just put 3 dots and say it is an object, for example:
```yaml
...
enable: <enable defines a boolean pointer whether we want the daemonset to exist or not>
podConfig:
  <same object with the same name in restic>
  ...
uploaderType: <The type of uploader to transfer the data of pod volumes, the supported values are 'restic' or 'kopia'>
```